service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
+	quitChan chan bool
+}
+
+// AssignHandlers creates and assigns the appropriate bridge
+// events to be listen for and intercepted on `mux` before forwarding
+// to `gcs` for handling.
+func (b *Bridge) AssignHandlers(mux *Mux, gcs core.Core) {
+	b.coreint = gcs
+	mux.HandleFunc(prot.ComputeSystemCreateV1, b.createContainer)
+	mux.HandleFunc(prot.ComputeSystemExecuteProcessV1, b.execProcess)
+	mux.HandleFunc(prot.ComputeSystemShutdownForcedV1, b.killContainer)
+	mux.HandleFunc(prot.ComputeSystemShutdownGracefulV1, b.shutdownContainer)
+	mux.HandleFunc(prot.ComputeSystemSignalProcessV1, b.signalProcess)
+	mux.HandleFunc(prot.ComputeSystemGetPropertiesV1, b.listProcesses)
+	mux.HandleFunc(prot.ComputeSystemWaitForProcessV1, b.waitOnProcess)
+	mux.HandleFunc(prot.ComputeSystemResizeConsoleV1, b.resizeConsole)
+	mux.HandleFunc(prot.ComputeSystemModifySettingsV1, b.modifySettings)
+}
+
+// ListenAndServe connects to the bridge transport, listens for
+// messages and dispatches the appropriate handlers to handle each
+// event in an asynchronous manner.
+func (b *Bridge) ListenAndServe() (conerr error) {
+	const commandPort uint32 = 0x40000000
+
+	var err error
+	b.commandConn, err = b.Transport.Dial(commandPort)
+	if err != nil {
+		return errors.Wrap(err, "bridge: failed creating the command Connection")
+	}
+	logrus.Info("bridge: successfully connected to the HCS via HyperV_Socket\n")
+
+	requestChan := make(chan *Request)
+	requestErrChan := make(chan error)
+	b.responseChan = make(chan bridgeResponse)
+	responseErrChan := make(chan error)
+	b.quitChan = make(chan bool)
+
+	defer close(requestChan)
+	defer close(requestErrChan)
+	defer close(b.responseChan)
+	defer close(responseErrChan)
+	defer close(b.quitChan)
+
+	// receive bridge requests and schedule them to be processed.
+	go func() {
+		for {
+			header := &prot.MessageHeader{}
+			if err := binary.Read(b.commandConn, binary.LittleEndian, header); err != nil {
+				requestErrChan <- errors.Wrap(err, "bridge: failed reading message header")
 			}
-		case prot.ComputeSystemWaitForProcessV1:
-			logrus.Info("received from HCS: ComputeSystemWaitForProcessV1")
-			response, err = b.waitOnProcess(message, header)
-			if err != nil {
-				logrus.Error(err)
-			} else {
-				// If no error occurred, don't respond until the process has
-				// exited.
-				response = nil
+			message := make([]byte, header.Size-prot.MessageHeaderSize)
+			if _, err := io.ReadFull(b.commandConn, message); err != nil {
+				requestErrChan <- errors.Wrap(err, "bridge: failed reading message payload")
 			}
-		case prot.ComputeSystemResizeConsoleV1:
-			logrus.Info("received from HCS: ComputeSystemResizeConsoleV1")
-			response, err = b.resizeConsole(message)
+			logrus.Infof("bridge: read message '%s'\n", message)
+			requestChan <- &Request{header, message}
+		}
+	}()
+	// process each bridge request async and create the response writer
+	go func() {
+		for req := range requestChan {
+			go func(r *Request) {
```

Comment:
Do you need to pass in `req` as a parameter, or can you just use the fact that `func` is a closure to use that variable?